`InoxRenderer` trait.

Richardn2002 opened the pull request:
GL renderer reimplemented using this trait.

The major part of this refactor is to finally resolve the mess of part/masks/composites being weaved together. The source of the mess is that drawables have the distinctions of `drawOne` and `drawSelf` (using ref impl terms) and "used or not, as a mask". The previous impl just blindly models the OOP pattern in the ref impl with two `bool`s being passed around. However the proper way is to factor out the commonalities instead of functions calling each other back and forth. 

Problems automatically emerged after this refactor:

0. recursive compositing just cannot happen. all checks are removed
1. find that the previous impl doesn't support simple mesh masks
2. find that the previous impl doesn't support composite node as mask
3. find that rendering to texture is not strictly enforced. in the framework there is easy way to customize the render flow
4. the data structure can be reworked a bit. after the refactor one will notice that some now separate `struct`s almost always appear together

Richardn2002 committed:
`InoxRenderer` trait.
GL renderer reimplemented using this trait.
The major part of this refactor is to finally resolve the mess of part/masks/composites being weaved together. The source of the mess is that drawables have the distinctions of `drawOne` and `drawSelf` (using ref impl terms) and "used or not, as a mask". The previous impl just blindly models the OOP pattern in the ref impl using with two `bool`s being passed around. However the proper way is to factor out the commonalities instead of functions calling each other back and forth.
Problems automatically emerged after this refactor:
0. recursive compositing just cannot happen. all checks are removed
1. find that the previous impl doesn't support simple mesh masks
2. find that the previous impl doesn't support composite node as mask
3. find that rendering to texture is not strictly enforced. in the framework there is easy way to customize the render flow
4. the data structure can be reworked a bit. after the refactor one will notice that some now separate `struct`s almost always appear together

Speykious commented on inox2d/src/render.rs:
```diff
@@ -218,3 +222,173 @@ impl Puppet {
+    fn end_composite(&self) -> Result<(), Self::Error>;
+
+    fn draw_composite_self(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error> {
+        self.begin_composite()?;
+
+        for &uuid in children {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            if let (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) =
+                (&node.data, &node_render_ctx.kind)
+            {
+                if as_mask {
+                    self.draw_part_self(true, camera, node_render_ctx, part, part_render_ctx)?;
+                } else {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+            } else {
+                // composite inside composite simply cannot happen
+            }
+        }
+
+        self.end_composite()?;
+        Ok(())
+    }
+
+    fn draw_composite(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        composite: &Composite,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
```
Why are `draw_part` and `draw_composite_self` the ones that have default implementations? Shouldn't it be the same kind of method across both, so `draw_part` and `draw_composite` or `draw_part_self` and `draw_composite_self`?

Also I think names like `draw_one_part` and `draw_one_composite` would be more explicit.

Richardn2002 replied:
These are the two considerations behind the designs of these 4 render functions:

0. `draw_part` and `draw_composite` are the top level draw functions, which when called (with `as_mask` set to `false`) on a fresh canvas, should yield one complete drawable.
1. `draw_part_self` and `draw_composite_self` are respectively, components that get called by `draw_part` `draw_composite`. They "belong" to the larger outside function, providing one step of the drawing process. In specific, they are named with `_self` because they draw the very content of the drawable in question"
    - For `Part`, `draw_part_self` draws the mesh warped texture of a `Part`.
    - For `Composite`, `draw_composite_self` draws all children content.

    However the outer part exist because only drawing the very content belongs to self does not complete the drawing. For `Part`, `Mask`s need to be deployed before drawing, and for `Composite`, composited content needs to be transferred to a correct location.

Then I can explain why certain functions are default, or to say I want to provide users with a ready-to-use fixed solution:

- For `Part`, as mask related functions are implemented elsewhere, and how `Mask` interact with `Part` is defined in the Inochi2D standard, we can just call them in a predefined order to deploy masks. What we don't know without renderer-specific knowledge is how to actually render a texture warped by mesh. We can almost say this step is the basic building block of all visual results. So `draw_part` is defined by us, calling `draw_part_self` and mask related functions, with `draw_part_self` implemented by the user.
- For `Composite`, the main content is then compositing all the content of its children, who are all `Part`s. However as all `Part` rendering is defined elsewhere, we already know how to draw all children. What we don't know without renderer-specific knowledge is how to actually transfer the composited content to the correct location. So `draw_composite_self` is defined by us, being called by `draw_composite`, with `draw_composite` implemented by the user.

Summarizing, we can give these 4 functions completely different names so that no one will be confused about one `_self` is default while the other one is not. Anyway the above is almost the explanation for how Inochi2D renders.

Speykious replied:
Hmmm, I see.

So about the suffixes, how about `draw_part_mesh` and `draw_composite_children` (both to replace `_self`)? It's slightly more descriptive (even though at the end of the day all these functions are gonna be documented).

Richardn2002 replied:
I have gone out of this mind game and figured out that the trick is not to try to give `Part` and `Composite` related methods similar names. Users will have to implement `draw_part_content` because we don't know how the rendering should be happening. Users also will have to implement `prepare_composite_draw` and `finish_composite_content` (these are the only two composite-related methods the user have to implement) because we don't know how the composite buffer is prepared and read after drawing is complete. Now the names are more descriptive and users won't mess them up.
And we do our job by calling the three methods in the right way. Users won't see how we name the internal functions anyway.

Speykious replied:
I see. I think I like this solution better. Though as a last suggestion, what about `begin_composite_content` as opposed to `prepare_composite_draw`? The pair of methods you have to implement would be the same and still distinct from `draw_part_content` so I think it's gonna be fine.

Speykious commented on inox2d/src/render.rs:
```diff
@@ -218,3 +222,173 @@ impl Puppet {
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
+
+    fn draw_drawables(&self, camera: &Mat4, puppet: &Puppet) -> Result<(), Self::Error> {
```
Maybe just `draw()` or `draw_puppet()`? `draw_drawables` feels weird and redundant.

Richardn2002 replied:
Naming of course can be changed. The only comment I want to make is that this call iterates over all the "top-level" drawables, which is defined at [source/inochi2d/core/puppet.d:212, branch v0_8](https://github.com/Inochi2D/inochi2d/blob/273dd40f2e406e9044ccfde2bf77999b15369b3b/source/inochi2d/core/puppet.d#L212C1-L218C22) as

```d
/**
    A list of parts that are not masked by other parts

    for Z sorting
*/
@Ignore
Node[] rootParts;
```

Consequentially speaking this call renders the whole puppet, so yeah maybe we can eventually use a simple name like `draw()`, and leave the drawable remarks to the comment.

Speykious replied:
Yeah. That it iterates over all the root parts and z-sorted is an implementation detail, it's better to be explicit about what it does in the end.

Richardn2002 replied:
I think we can one day ask Luna about this. From the actual implementations of the ref impl `rootParts` seems like just simply the `top-level` `Part`s and `Composite`s, excluding `Mask`s and `Composite` children. This set has nothing to do with "not masked by others". This makes sense to me, and the current inox impl is doing this...

...almost. I now find a bug, where composite children are present in the `nodes_zsorted`, thus `drawables_zsorted`. The render seems to work now because Aka chan is not that complex, but this is definitely wrong: A composite children will be drawn during the composite draw, and will be drawn again independently as a normal `Part`.

Speykious commented on inox2d/src/render.rs:
```diff
@@ -218,3 +222,173 @@ impl Puppet {
+                } else {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+            } else {
+                // composite inside composite simply cannot happen
+            }
+        }
+
+        self.end_composite()?;
+        Ok(())
+    }
+
+    fn draw_composite(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        composite: &Composite,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
+
+    fn draw_drawables(&self, camera: &Mat4, puppet: &Puppet) -> Result<(), Self::Error> {
+        for &uuid in &puppet.render_ctx.drawables_zsorted {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            match (&node.data, &node_render_ctx.kind) {
+                (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) => {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+
+                (InoxData::Composite(ref composite), RenderCtxKind::Composite(ref children)) => {
+                    self.draw_composite(false, camera, composite, puppet, children)?;
+                }
+
+                _ => {
+                    // This clearly is sign that the data structure needs rework
+                    todo!();
+                }
+            }
```
I think we can manage without for now. Maybe this match statement could be its own method from the `InoxRenderer` trait with this as the default implementation, something like `fn draw_node(&self, node: &InoxNode, node_render_ctx: &RenderCtx, camera: &Mat4, puppet: &Puppet)`?

Though the more I look at this, the more I feel like a way to make this flexible for any kind of node variant or custom type would be to turn nodes into a trait again and abstract the drawing process there... But I don't know if it's a good idea.

Richardn2002 replied:
I suggest rework because `PartRenderCtx` info is always coupled with `Part` and `CompositeRenderCtx` is always coupled with `Composite`. Because of we insisted on separating them while they are actually coupled, we are now left with the `_ => ()` branch, signaling two meaningless combinations.

I am suggesting putting `RenderCtx` info into `Part` and `Composite`. They are now separated because in the renderless API refactor by me I factored out all info directly used for rendering. But we are inox2d and every data structure is directly or indirectly related to the render of a puppet, so really no point to make the distinction. And indexing by `uuid` twice to retrieve two pieces of info takes twice the overhead.

By the way, please note that in Rust `enum`s are implemented as tagged unions, so in future refactors please mind not to make different variants of an enum taking too different spaces or we will be wasting a lot space on padding.

Richardn2002 replied:
Oh I forgot to reply to the `draw_node` part: The thought of factoring out this `match` appeared again and again through out the development process, however the two places that deploy this match have slight differences which make such a function useless (you will need two `bool`s for this function). The conclusion I reached is that it is because the currently segmented data structure, a simple access to drawable data from one node is tedious so that one will try to factor this process out. But if the data structure is reworked maybe this will no longer be the case.

Speykious replied:
> But we are inox2d and every data structure is directly or indirectly related to the render of a puppet, so really no point to make the distinction.

I guess we can trade off some amount of memory in nodes and gain back some inside the renderer since we won't need the hashmap anymore. You're not wrong, trying to separate the two has negligible benefits given the use-case of inox2d.



> By the way, please note that in Rust `enum`s are implemented as tagged unions, so in future refactors please mind not to make different variants of an enum taking too different spaces or we will be wasting a lot space on padding.

Don't worry, I always keep that in mind. :p

Speykious commented on inox2d/src/render.rs:
```diff
@@ -218,3 +222,173 @@ impl Puppet {
+    type Context;
+    type Error;
+
+    // create a new renderer, given rendering context
+    fn new(ctx: Self::Context) -> Result<Self, Self::Error>;
+
+    // do any model-specific setups, e.g. creating buffers with specific sizes
+    // after this step, the model provided should be renderable
+    fn prepare(&mut self, model: &Model) -> Result<(), Self::Error>;
+
+    // resize viewport
+    fn resize(&mut self, w: u32, h: u32) -> Result<(), Self::Error>;
```
You can actually make all of these comments into documentation by using `///` instead of `//`. :D

Richardn2002 replied:
:D

(well, where is the official documentation on when and how to use `///` and `//`. I really don't know.

---EDIT---
https://doc.rust-lang.org/reference/comments.html

Speykious commented on inox2d/src/render.rs:
```diff
@@ -218,3 +222,173 @@ impl Puppet {
     }
 }
+
+pub trait InoxRenderer
+where
+    Self: Sized,
+{
+    type Context;
+    type Error;
```
These types need a bit of documentation like the rest of the methods. I think a small doctype for the trait itself would be nice too.

Richardn2002 replied:
Speaking of this, do you think making almost every function returning `Result<(), Self::Error>` a good idea. Basically I am trying to abstract outcomes of one render step without knowledge specific to the renderer. The only outcome other then graphical outputs is failures when I think of it. So I let users define a custom `Error` type.

And `Context` is apparently render context. GL renderers can put `glow::Context` there, and `wgpu` renderers can put their beloved

```rust
struct State {
    surface: wgpu::Surface,
    device: wgpu::Device,
    queue: wgpu::Queue,
    config: wgpu::SurfaceConfiguration,
    size: winit::dpi::PhysicalSize<u32>,
    window: Window,
}
```

in.

Speykious replied:
Ultimately, the more I think about it, the more I think that anything that pertains to rendering or doing something in a loop should just return `()` (i.e. nothing).

So the only function that is worth returning a `Result<(), Self::Error>` imo is `prepare`.
I would've said `new` as well, except I just realized I don't think it needs to be in the trait. The trait doesn't need to manage how an Inox Renderer should be constructed, only how it behaves after construction.

Richardn2002 replied:
`new` has to be in the trait, with the `type Context` defined by the renderer implementation just to make it work, so that we can ensure the main code works no matter which implementation is selected. If `new` has different signatures some confusions may happen.

Speykious replied:
What I'm saying is that the main code should not control how the renderer is constructed, it's not necessary. So that means that the `Context` GAT is also not necessary.

It will work regardless of the renderer, but we won't need inox2d to construct the renderer itself for it to be renderer-agnostic.

Speykious reviewed (changes requested):
I didn't review the OpenGL implementation code changes yet, but as for the abstraction itself, I like where this is going!
I appreciate that you documented all these functions. Though I have a few questions and suggestions for change.

Speykious commented:
And of course once all of this is good we have to make sure that the WebGL example still compiles

Richardn2002 commented:
> And of course once all of this is good we have to make sure that the WebGL example still compiles

It does! The example runs fine just as before, however with

```
ERROR opengl: Error: GL_INVALID_OPERATION in glFramebufferTexture2D(non-existent texture 1)
ERROR opengl: Error: GL_INVALID_OPERATION in glFramebufferTexture2D(non-existent texture 2)
ERROR opengl: Error: GL_INVALID_OPERATION in glFramebufferTexture2D(non-existent texture 3)
ERROR opengl: Error: GL_INVALID_OPERATION in glFramebufferTexture2D(non-existent texture 4)
```

during renderer initialization.

Well we all know in many scenarios GL just continues to work with a lot GL errors, but it would be nice to fix this.

--- EDIT ---

Oh shoot, you say `WebGL`. I misread and thought about `OpenGL`. Yeah the `WebGL` example failed.

Richardn2002 committed:
GL_INVALID_OPERATION in glFramebufferTexture2D resolved
previously in `new`ing the renderer a `viewport` is passed in to resize the viewport
I find this annoying, and decided that one should always `new` and then `.resize`
however there is a left over code in new related to texture reattaching

Richardn2002 committed:
`InoxRendererCommon` trait.
Three methods that shouldn't be overridden by the user.

Richardn2002 committed:
Fixed WebGL example
Use glow::VertexArray abstraction instead of glow::NativeVertexArray for webgl portability

Richardn2002 committed:
Tweak composite-related methods definition.
Now the user will be less confused, and function signatures indeed shrink.

Richardn2002 committed:
Remove most `-> Result<(), Self::Error>`
and fixing function comments using `///`

Richardn2002 commented:
CI failed due to 404. Nice work GitHub!

Speykious committed:
Comment out wasm32 target setting

Speykious committed:
Remove `new` method from `InoxRenderer`

It doesn't need to be responsible for the renderer's initialization.

Richardn2002 committed:
Improve node tree overall usage.
`zsorted_children()` no longer returns parent
comments added
during constructing `RenderCtx`, making clear that some procedures require zsortedness and exclusion of certain nodes, while some procedures just require all nodes in arbitary order
Use HashMap instead of BTreeMap for the uuid -> index map. No point sorting uuids.

Richardn2002 committed:
Fix WebGL example.

Speykious committed:
Polish doc comments

There might be more documentation work needed later on though.

Speykious committed:
apt update (maybe CI won't fail anymore)

Speykious reviewed (approved):
LGTM 👍
